torch/csrc/cuda/nccl.cpp

```diff
@@ -807,6 +807,47 @@ void recv(
 #endif
 }
 
+
+void gather(
+    const at::Tensor& inputs,
+    at::Tensor& outputs,
+    ncclComm_t _comm,
+    at::cuda::CUDAStream& stream,
+    int32_t root) {
+#ifdef USE_NCCL
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR == 2) && (NCCL_MAJOR * 10 + NCCL_MINOR) >= 27
+  using namespace torch::cuda::nccl::detail;
+
+  auto comm = to_nccl_comm(_comm);
+  int numranks, cur_rank;
+  NCCL_CHECK(ncclCommCount(comm, &numranks));
+  NCCL_CHECK(ncclCommUserRank(comm, &cur_rank));
+
+  size_t count = inputs.numel();
+  auto type = to_nccl_data_type(inputs);
+  const auto* sendbuff = reinterpret_cast<char*>(inputs.data_ptr());
+
+  NCCL_CHECK(ncclGroupStart());
+  if (cur_rank == root)
+  {
+    for (int r = 0; r < numranks; r++)
+    {
+      auto* recvbuff =  reinterpret_cast<char*>(outputs[r].data_ptr());
+      NCCL_CHECK(ncclRecv(recvbuff, count, type, r, comm, stream));
+    }
+  }
+  NCCL_CHECK(ncclSend(sendbuff, count, type, root, comm, stream));
```

Comment:
nit: Shouldn't we have the send before the recv? I don't think it has any correctness issues since NCCL would handle it, but it is awkward that the root rank calls recv before send.